flexible-event/index.js

```diff
@@ -0,0 +1,253 @@
+ * @param flipProbability - the channel keeps the input the same with probability
+ *   1 - flipProbability, and flips the input to one of the other q - 1
+ *   symbols (uniformly) with the remaining probability of flipProbability.
+ *
+ * @returns - The capacity of the q-ary symmetric channel for given flipProbability.
+ *   In general, the capacity is defined as the maximum, over all input
+ *   distributions, of the mutual information between the input and output of
+ *   the channel. In the special case of the q-ary symmetric channel, a
+ *   closed-form expression is known, which we use here.
+ */
+function capacityQarySymmetricChannel(log2q, flipProbability) {
+  return log2q - h(flipProbability) - flipProbability * Math.log2(Math.pow(2, log2q) - 1)
+}
+
+
+// Returns the maximum information for a source using the flexible event API.
+function maxInformationGain(numStates, epsilon) {
+  const flipProb = flipProbabilityDp(numStates, epsilon)
+  return capacityQarySymmetricChannel(Math.log2(numStates),
+      flipProb * (numStates - 1) / numStates)
+}
+
+// Returns the effective epsilon and flip probability needed to satisfy an information gain bound
+// given a number of output states in the q-ary symmetric channel.
+function epsilonToBoundInfoGainAndDp(
+    numStates,
+    infoGainUpperBound,
+    epsilonUpperBound,
+    tolerance = 0.00001) {
+  // Just perform a simple binary search over values of epsilon.
+  let epsLow = 0
+  let epsHigh = epsilonUpperBound
+
+  while (true) {
+    const epsilon = (epsHigh + epsLow) / 2
+    const infoGain = maxInformationGain(numStates, epsilon)
+
+    if (infoGain > infoGainUpperBound) {
+      epsHigh = epsilon
+      continue
+    }
+
+    // Allow slack by returning something slightly non-optimal (governed by the tolerance)
+    // that still meets the privacy bar. If epsHigh == epsLow we're now governed by the epsilon
+    // bound and can return.
+    if (infoGain < infoGainUpperBound - tolerance && epsHigh != epsLow) {
+      epsLow = epsilon
+      continue
+    }
+
+    return [epsilon, flipProbabilityDp(numStates, epsilon)]
+  }
+}
+
+function printConfigData(config, epsilon, sourceType) {
+  const numStates = numFlexibleStates(config)
+  const infoGain = maxInformationGain(numStates, epsilon)
+  const flipProb = flipProbabilityDp(numStates, epsilon)
+
+  console.log(`Number of possible different output states: ${numStates}`)
+  console.log(`Information gain: ${infoGain} bits`)
+  console.log(`Flip percent: ${100 * flipProb}%`)
+
+  const infoGainDefault = maxInformationGain(
+        numFlexibleStates(DEFAULT_CONFIG[sourceType]), epsilon)
+
+  if (infoGain > infoGainDefault) {
+      const [newEps, flipProb] = epsilonToBoundInfoGainAndDp(numStates, infoGainDefault, epsilon)
+      console.log(
+          `WARNING: info gain of ${infoGain} > ${infoGainDefault} for ${sourceType} sources. Would require a ${100 * flipProb}% flip chance (effective epsilon = ${newEps}) to resolve.`)
```

Comment:
The output is a lot nicer to read if these are capped to a certain # of decimal points using e.g. `toFixed(<decimal places>)`